pwnlib/util/misc.py

```diff
@@ -270,6 +271,69 @@ def mkdir_p(path):
         else:
             raise
 
+def sh_prepare(var, s, export = False):
+    """Outputs a posix compliant shell command that will put `s` into the
+    variable named `var`. `s` cannot contain null-bytes.
+
+    It is assumed that `var` is a valid name for a variable in the shell.
+
+    Examples:
+
+        >>> print sh_prepare('X', 'foobar')
+        X=foobar
+        >>> print sh_prepare('X', 'foo bar')
+        X='foo bar'
+        >>> print sh_prepare('X', "foo'bar")
+        X="foo'bar"
+        >>> print sh_prepare('X', "foo\\\\bar")
+        X='foo\\bar'
+        >>> print sh_prepare('X', "foo\\\\'bar")
+        X="foo\\\\'bar"
+        >>> print sh_prepare('X', "foo\\x01'bar")
+        X=$(printf 'foo\\001\\047barx');X=${X%x}
+        >>> print sh_prepare('X', "foo\\x01'bar", export = True)
+        X=$(printf 'foo\\001\\047barx');export X=${X%x}
+        >>> print `subprocess.check_output('%s;echo -n "$X"' % sh_prepare('X', "foo\\x01'bar"), shell = True)`
+        "foo\\x01'bar"
+    """
+
+    if '\x00' in s:
+        log.error("sh_prepare(): Cannot create a null-byte")
+
+    very_good = set(string.ascii_letters + string.digits)
+    good      = (very_good | set(string.punctuation + ' ')) - set("'")
+    alt_good  = (very_good | set(string.punctuation + ' ')) - set('!')
+
+    if '\x00' in s:
+        log.error("sh_string(): Cannot create a null-byte")
+
+    export = 'export ' if export else ''
+
+    if all(c in very_good for c in s):
```

Comment:
I feel like we should be able to skip almost all of this logic and just use sh_string for both.  This is basically duplicating the same logic.

sh-3.2$ export $(echo hello)=$(echo world)
sh-3.2$ echo $hello
world